⚙️[WIP] Adding object effects manager in runtime.

HarsimranVirk opened the pull request:
I'm not sure exactly what happened with the previous branch 😬 (maybe too much merging?). I've cherry-picked my commits from the previous branch to this branch, should be good now.

TODO:
- [x] Rework the "layer" passed to effect to only pass a few useful stuff like the size of the area (either the size of the layer or the object) and the elapsed time?
- [x] Handle hot-reloading:
  - [x] Adding a new effect should work
  - [x] Removing an effect should remove it
  - [x] Changing effect parameters should update it.
- [x] Test that actions to enable/disable an effect are working
- [ ] Add some tests to create a runtime object, add an effect, change its parameters, delete the effect.

HarsimranVirk commented on Core/GDCore/Events/CodeGeneration/EffectsCodeGenerator.cpp:
```diff
@@ -34,6 +35,15 @@ void ExposeProjectEffects(
     }
   }
+
+  // Add object effects
+  for (std::size_t s = 0; s < project.GetObjectsCount(); s++) {
+    auto& effects = project.GetObject(s).GetEffects();
+    for (std::size_t e = 0; e < effects.GetEffectsCount(); e++) {
+      auto& effect = effects.GetEffect(e);
+      worker(effect);
+    }
+  }
```
@4ian this doesn't seem to include (and register) effects. Is something missing here? I'm not sure about what has to be done here.

4ian replied:
This is only for global objects (i.e: project objects). You also must do this for every layout objects :)

HarsimranVirk commented on GDJS/Runtime/pixi-renderers/pixi-filters-tools.ts:
```diff
@@ -88,7 +88,7 @@ namespace gdjs {
     export type FilterCreator = {
       /** Function to call to create the filter */
-      makePIXIFilter: (layer: gdjs.Layer, effectData: EffectData) => any;
+      makePIXIFilter: (layer: gdjs.Layer | gdjs.RuntimeObject, effectData: EffectData) => any;
```
I guess this is def wrong, I'll check each effect later, to be sure that there are no errors.

4ian replied:
Yeah we probably actual need to either remove this first argument from all filters, or rename it according to what it's used for... no big deal for now.

4ian reviewed (commented):
Seems like a good start :) I think if you add the missing object effects in `ExposeProjectEffects`, you should see runtime objects with effects properly applied.

We can then work on:
- extending the PixiObjectEffectsManager interface to support adding/removing/modifying effects. 
- we should call `update` on effects too.
- expose actions/conditions to manipulate effects like for layers

👍

HarsimranVirk commented:
@4ian I'm sorry for the delay, I had a lot of work the past couple of weeks 😅.

> we should call update on effects too.

As I understand, all of the derived classes of the `RuntimeObject` have their own implementations of the `update()` method, is there a way to update the filters from the base class itself?

HarsimranVirk commented:
@4ian I've added the methods related to effects, (please check if the design makes sense), but I have no idea about how to expose them as actions or conditions (I know how to do it for extensions, but these methods are defined on `RuntimeObject` itself). Can you explain how to do this?

Bouh commented:
> @4ian I've added the methods related to effects, (please check if the design makes sense), but I have no idea about how to expose them as actions or conditions (I know how to do it for extensions, but these methods are defined on `RuntimeObject` itself). Can you explain how to do this?

1) A part of [the declaration](https://github.com/4ian/GDevelop/blob/master/Core/GDCore/Extensions/Builtin/BaseObjectExtension.cpp) is done in C++
2) [Binding is done in C++](https://github.com/4ian/GDevelop/blob/master/GDCpp/GDCpp/Extensions/Builtin/BaseObjectExtension.cpp) to handle the JS part
3) An implementation is done in JS (you have already done that part)

4ian commented on Core/GDCore/Extensions/Builtin/BaseObjectExtension.cpp:
```diff
@@ -1373,6 +1373,94 @@ void GD_CORE_API BuiltinExtensionsImplementer::ImplementsBaseObjectExtension(
+  obj.AddAction("RemoveEffect",
+                _("Remove an effect"),
+                _("Remove an effect from the object"),
+                _("Remove _PARAM1_ effect from the _PARAM0_"),
```
Remove effect _PARAM1_ from _PARAM0_

(never "the" before object names)

4ian commented on Core/GDCore/Extensions/Builtin/BaseObjectExtension.cpp:
```diff
@@ -1373,6 +1373,94 @@ void GD_CORE_API BuiltinExtensionsImplementer::ImplementsBaseObjectExtension(
+  
+  obj.AddAction("RemoveEffect",
+                _("Remove an effect"),
+                _("Remove an effect from the object"),
```
Period at the end of all descriptions.

4ian commented on Core/GDCore/Extensions/Builtin/BaseObjectExtension.cpp:
```diff
@@ -1373,6 +1373,94 @@ void GD_CORE_API BuiltinExtensionsImplementer::ImplementsBaseObjectExtension(
+  obj.AddAction("EnableEffect",
+                _("Enable an object effect"),
+                _("Enable an effect on the object"),
+                _("Enable _PARAM1_ effect on _PARAM0_ : _PARAM2_"),
```
`Enable effect _PARAM1_ on _PARAM0_: _PARAM2_`

4ian commented on Core/GDCore/Extensions/Builtin/BaseObjectExtension.cpp:
```diff
@@ -1373,6 +1373,94 @@ void GD_CORE_API BuiltinExtensionsImplementer::ImplementsBaseObjectExtension(
+      .MarkAsSimple();
+  
+  obj.AddAction("SetEffectDoubleParameter",
+                _("Set a double parameter for effect"),
```
Look at how it was done for layers:
`Effect parameter (number)`

4ian commented on Core/GDCore/Extensions/Builtin/BaseObjectExtension.cpp:
```diff
@@ -1373,6 +1373,94 @@ void GD_CORE_API BuiltinExtensionsImplementer::ImplementsBaseObjectExtension(
+  
+  obj.AddAction("SetEffectDoubleParameter",
+                _("Set a double parameter for effect"),
+                _("Set a double parameter for an effect of the object"),
```
Same, copy the exact same descriptions and names as for layers:
```suggestion
                _("Change the value of a parameter of an effect.") + "\n" +
              _("You can find the parameter names (and change the effect "
                "names) in the effects window."),
```

4ian commented on Core/GDCore/Extensions/Builtin/BaseObjectExtension.cpp:
```diff
@@ -1373,6 +1373,94 @@ void GD_CORE_API BuiltinExtensionsImplementer::ImplementsBaseObjectExtension(
+  obj.AddAction("SetEffectDoubleParameter",
+                _("Set a double parameter for effect"),
+                _("Set a double parameter for an effect of the object"),
+                _("Set the value of _PARAM2_ as _PARAM3_ for the " 
+                  "effect _PARAM1_ of object _PARAM0_"),
```
`Set _PARAM2_ to _PARAM3_ for effect _PARAM3_ of _PARAM0_`

4ian commented on Core/GDCore/Extensions/Builtin/BaseObjectExtension.cpp:
```diff
@@ -1373,6 +1373,94 @@ void GD_CORE_API BuiltinExtensionsImplementer::ImplementsBaseObjectExtension(
+                "res/actions/effect24.png",
+                "res/actions/effect.png")
+      .AddParameter("object", _("Object"))
+      .AddParameter("string", _("Effect Name"))
+      .AddParameter("string", _("Parameter Name"))
+      .AddParameter("number", _("Parameter Value"))
```
Follow the exact same namings as for layers:
```suggestion
      .AddParameter("string", _("Effect"))
      .AddParameter("string", _("Parameter name"))
      .AddParameter("expression", _("New value"))
```

4ian commented on Core/GDCore/Extensions/Builtin/BaseObjectExtension.cpp:
```diff
@@ -1373,6 +1373,94 @@ void GD_CORE_API BuiltinExtensionsImplementer::ImplementsBaseObjectExtension(
+      .AddParameter("number", _("Parameter Value"))
+      .MarkAsSimple();
+  
+  obj.AddAction("SetEffectStringParameter",
```
Same remarks for the next twos for the title/description/sentence/parameters :)

4ian commented on Core/GDCore/Extensions/Builtin/BaseObjectExtension.cpp:
```diff
@@ -1373,6 +1373,94 @@ void GD_CORE_API BuiltinExtensionsImplementer::ImplementsBaseObjectExtension(
+      .MarkAsSimple();
+  
+  obj.AddCondition("IsEffectEnabled",
+                  _("Effect Enabled"),
+                  _("Check if effect is enabled"),
+                  _("_PARAM1_ is enabled on _PARAM0_"),
```
```suggestion
                  _("Effect is enabled"),
                    _("The effect on an object is enabled"),
                    _("Effect _PARAM2_ on _PARAM1_ is enabled"),
```

4ian commented on Core/GDCore/Extensions/Builtin/BaseObjectExtension.cpp:
```diff
@@ -1373,6 +1373,94 @@ void GD_CORE_API BuiltinExtensionsImplementer::ImplementsBaseObjectExtension(
+      .AddParameter("string", _("Effect Name"))
+      .MarkAsSimple();
+
+  obj.AddCondition("HasEffect",
```
Let's remove this one, effects can't be added/removed at runtime, so we should just have a condition to check if an effect is enabled, not if it is present.

4ian commented on GDJS/Runtime/pixi-renderers/pixi-filters-tools.ts:
```diff
@@ -14,7 +14,7 @@ namespace gdjs {
     };
 
-    const _filterCreators: {
+    export const _filterCreators: {
```
Seems weird to export something that has an underscore? Either we need to rename this, or use something else.

HarsimranVirk replied:
I think I checked it out by mistake (I was trying to figure out what are these objects exactly meant for). My bad, I'll fix it. 😬

4ian commented on GDJS/Runtime/pixi-renderers/pixi-object-effects-manager.ts:
```diff
@@ -0,0 +1,208 @@
+    update(runtimeObject: RuntimeObject, layer: Layer) {
+      const filters: Hashtable<
+        PixiFiltersTools.Filter
+      > | void = this._filters.get(runtimeObject.getName());
```
undefined, not void.

4ian commented on GDJS/Runtime/pixi-renderers/pixi-object-effects-manager.ts:
```diff
@@ -0,0 +1,208 @@
+      const renderer = runtimeObject.getRendererObject();
+      renderer.filters = (renderer.filters || []).concat(filter.pixiFilter);
+
+      if (!this._filters.containsKey(runtimeObject.getName())) {
```
This approach unfortunately won't really work, because you're storing filters for objects according to their name, which has potential issues if we later want to store more information differing for each instance. Also, it's going to be quite slow, at least suboptimal, to use a hashtable to get the filters everytime we need it for an object, having to read the object name and use it as a key.

Instead, each gdjs.RuntimeObject must have its `_filters`. Much like every layer has it :)

4ian replied:
It's still fine I think, because we consider `_filters` as an "opaque" data structure which is handled by `gdjs.PixiObjectEffectsManager`. 
In a gdjs.RuntimeObject, there is already usually a `_renderer` object (depending on the object, but it's always and by convention called like this). We now add a `_filters`, to gdjs.RuntimeObject, which has a structure that is both known (it's a map to the current filters applied to the object) but that also contain the "renderer specific filters", aka here the "PIXI filters". 
If we were to implement this for another renderer, like Cocos, we would still have it, but we would have a `gdjs.CocosObjectEffectsManager` that would store cocos effects inside.

4ian commented on GDJS/Runtime/pixi-renderers/pixi-object-effects-manager.ts:
```diff
@@ -0,0 +1,208 @@
+    addEffect(
+      runtimeObject: RuntimeObject,
+      effectData: EffectData,
+      layer: Layer
```
Don't pass the layer to addEffect or anywhere. Effects on objects should have no access to the layer at all.

HarsimranVirk replied:
I was passing the layer because, the `update` function of the filters require `getElapsedTime()`, for example here:
https://github.com/4ian/GDevelop/blob/f3a49ad2cfd48efa6bed17565041094390a199b6/Extensions/Effects/glitch-pixi-filter.ts#L10-L18
The `getElapsedTime()` of `runtime Object` basically calls the function of  `layer`:
https://github.com/4ian/GDevelop/blob/f3a49ad2cfd48efa6bed17565041094390a199b6/GDJS/Runtime/runtimeobject.ts#L193-L197

4ian replied:
In the future, maybe some objects will have a different time scale than other, so it's better to use the function provided by the object :)

Consider the "layer" object passed to the effects as a new type of object, which should contain basically "getElapsedTime, getWidth, getHeight". Not sure how to call it in TypeScript and as the parameter, but it's probably something we can call "EffectTarget"/"effectTarget".

HarsimranVirk replied:
Right, I think for the moment it can be a union type of gdjs.Layer and gdjs.RuntimeObject, but again, the `getElapsedTime()` 
 method is not equivalent at the moment for layers and runtimeobject (there's a difference of arguments). We might need some refactoring here.

4ian replied:
Ah right! Mmmm let's make this a TODO. We can think about changing what we pass to effects (rather than passing an entire layer or runtimeobject, we can pass a few useful functions, we'll see!)

Bouh commented:
Let me know when there is something to test in the application.

4ian commented:
Just to give a bit more explanation about `pixi-object-effects-manager.ts` => in this class, you should only have things that deal with Pixi filters. 
The `gdjs.RuntimeObject` should have, in its method related to effects (`setEffectDoubleParameter`, `enableEffect`, `addEffect`, etc...):
- manipulation, as needed, of `this._filters` (the filters applied on the object, but this could work with anything, not only PixiJS).
- manipulation, using gdjs.ObjectEffectsManager, of PIXI filters (by using methods of `gdjs.ObjectEffectsManager` that could be implemented using another rendering engine in the future).

In other words, just use `gdjs.ObjectEffectsManager` to do the PIXI specific stuff :) And store in `this._filters` the rest.
Give it a try and let me know!

4ian commented on GDJS/Runtime/pixi-renderers/pixi-object-effects-manager.ts:
```diff
@@ -0,0 +1,169 @@
+      renderer.filters = (renderer.filters || []).filter(
+        (pixiFilter) => pixiFilter !== filter.pixiFilter
+      );
+      delete filter[effectName];
```
You're sure about this line? `filter` is already a filter object, it does not have any `effectName` property.
You probably want to run delete on `runtimeObject.getFilters()`

HarsimranVirk replied:
Oh yes, right. I think at this point I should be adding some tests to avoid these silly mistakes. 😬

4ian replied:
Might be a good idea to think about unit/integration tests (probably integration tests if this requires PixiJS) yes, to check the good working of methods that add/change/delete effects :)

4ian commented on GDJS/Runtime/runtimeobject.ts:
```diff
@@ -70,6 +70,10 @@ namespace gdjs {
 
+    //Effects:
+    protected _effects: EffectData[];
+    protected _filters: Record<string, PixiFiltersTools.Filter> = {}; 
```
Note: we have here "PixiFiltersTools.Filter", which is PixiJS specific, but there is nothing inside the runtime object actually PixiJS specific, as everything is going through the object effects manager. So this is good! And we can always change the typing later without breaking anything.

4ian replied:
We might just want to rename this to `_rendererEffects`. Because "filters" is a PixiJS specific term, while here it's a table that contains the effects of the renderer (i.e: the effects of PIXI, i.e: the PIXI filters).
Other renderers in the future might have a different name ("shaders", etc.., so keeping the GDevelop term ("effect") is in theory better)

4ian commented:
Starting to look good! 
I'll be editing your first message with the list of things to do, like the question of the layer, hot-reloading (feel free to add/check stuff).

In terms of testing this for real, how far are we? Are you able to add an effect to an object and to see it on screen?

HarsimranVirk commented:
> In terms of testing this for real, how far are we? Are you able to add an effect to an object and to see it on screen?

Yes, we can enable/disable the effects on runtime using actions and conditions as well. I'm using the effects on `Player` object of `Platformer` game to test this. I'm able to change effects (enable/disable) based on the state of `Player` object (If it's jumping then apply a specific set of effects, if not apply a different set of effects, etc).

4ian commented on GDJS/Runtime/runtimeobject.ts:
```diff
@@ -730,6 +745,145 @@ namespace gdjs {
+     * Add a new effect, or replace the one with the same name.
+     * @param effectData The data of the effect to add.
+     */
+    addEffect(effectData) {
```
Missing typing: EffectData

4ian commented on GDJS/Runtime/runtimeobject.ts:
```diff
@@ -730,6 +745,145 @@ namespace gdjs {
+     * Remove the effect with the specified name
+     * @param effectName The name of the effect.
+     */
+    removeEffect(effectName) {
```
string

4ian commented on GDJS/Runtime/runtimeobject.ts:
```diff
@@ -98,6 +102,7 @@ namespace gdjs {
       );
       this._averageForce = new gdjs.Force(0, 0, 0);
       this._behaviorsTable = new Hashtable();
+      this._effects = objectData.effects;
```
You probably don't need and don't want to store effects from objectData.
These are the effects as they should be initially.

If you store this on all the objects, this means:
- all objects have a pointer to these, which consume memory.
- if I update the object data, thanks to the hot reload, all the objects are now having a pointer to the old effects data, which is a memory leak and outdated information.

This should actually not be needed. You only need this when you initialize the object, so just iterate on `objectData.effects`. The runtime effects are then stored in `_rendererEffects` (which contains the actual Pixi filters).

4ian commented:
> > In terms of testing this for real, how far are we? Are you able to add an effect to an object and to see it on screen?
> 
> Yes, we can enable/disable the effects on runtime using actions and conditions as well. I'm using the effects on `Player` object of `Platformer` game to test this. I'm able to change effects (enable/disable) based on the state of `Player` object (If it's jumping then apply a specific set of effects, if not apply a different set of effects, etc).

Great! Going in the right direction :) 
Let's go for tests, then hot reloading (with a new `_hotReloadRuntimeObjectsEffects(oldEffectData, newEffectData, runtimeObjects` method) and we'll see how we rework layer/objects passed to effects at the end.

Bouh commented:
I'll test it tonight :D

HarsimranVirk commented:
1. Added hot-reloading of effects.
2. Removed all the `runtimeObject` args from the methods of `ObjectEffectsManager` class, we don't need to pass the entire runtime object.

Remaining:

- [ ] Remove dependency of `Layer`.
- [ ] Add tests.

Bouh commented:
Good point there is no error on the compilation and at the opening 👍 

- The effects tab is present, it displays a hint, I think we can do like for the behaviors, an empty state instead of a hint.
![image](https://user-images.githubusercontent.com/1670670/125862204-f3eed18a-ad51-42d8-b602-9e7def36a4db.png)

- I recommend using the dropdown items list from the behaviors list into the effects list, there is a lot of parameters on effects as behaviors.
![image](https://user-images.githubusercontent.com/1670670/125862428-59945469-26de-4245-9519-c8e6b2c8ded5.png)

- With a Kawase (blur) effect on the object flicker. (See in video bellow) I suppose an issue with subpixels, it happened on other filters like Beveled edges.
- With a Kawase (blur) effect, when you move the object across the scene and off-screen,
the object becomes blurry when it approaches the window's borders. (See in video below)

https://user-images.githubusercontent.com/1670670/125863915-0b2b9000-3dd0-46a5-84e0-2ef6f543fcb7.mp4

- Displacement map, and Twist has as reference the world, not the object, not sure?
- Glitch effect interacts with the game border, I guess the referential is not correct for the effect, can you test it?


It looks promising! Some effects require to be adapted for objects.
The problem of precisions or subpixels I don't know what it is is very annoying (See video below)


https://user-images.githubusercontent.com/1670670/125866011-76efc205-ba7c-4143-917f-daa81ed915cf.mp4


https://user-images.githubusercontent.com/1670670/125866014-d618261d-c2e8-4e3c-8dae-72ef196a61e8.mp4

4ian commented:
Thanks for checking @Bouh 👍

For the first two remarks, I think it's a question of the design of the effects list that can be handled in another PR (and could even be done before that PR).

Good spot for the effects having issues. I think we need to be sure that all effects are adapted so that they are using the objects size/position for their effect (for example, a "twist" that is at position 0,5;0,5 must be centered on the object, not on the screen). Kawase is maybe not rendering properly because using the layer coordinates instead of the object?

> Displacement map, and Twist has as reference the world, not the object, not sure?

Yes, all effects must be checked to adapt them to use either the layer or the object (they should not even know, they should just receive the positions of the "target" - the target being the object or the layer).

HarsimranVirk commented:
@4ian how can we handle the case of `getElapsedTime()` of rutime object, as it's not equivalent to the layer's method?

> Yes, all effects must be checked to adapt them to use either the layer or the object (they should not even know, they should just receive the positions of the "target"

For this, I was thinking about using an interface `EffectsTarget`, which will be implemented by both the classes (RuntimeObject and Layer), and the "target" will simply be an implementation of `EffectsTarget`. But again, the problem here is with the `getElapsedTime()` method as the signature isn't the same. What refactoring should we do here? (Add a runtimeScene arg in layer's impl, or remove it from runtime object's impl, I'm not sure what should be the correct way.)

4ian commented on GDJS/Runtime/runtimeobject.ts:
```diff
@@ -123,6 +130,16 @@ namespace gdjs {
         this._behaviors[i].onCreated();
       }
+
+      const objectEffectsManager = this._runtimeScene
+        .getGame()
+        .getObjectEffectsManager();
+      for (const filterName in this._rendererEffects) {
+        objectEffectsManager.applyEffect(
+          this.getRendererObject(),
+          this._rendererEffects[filterName]
+        );
+      }
```
Can this be moved to the constructor? Or at least before the `onCreated` called on behaviors.
But 99% sure this should be in the constructor?

4ian replied:
Ah right this can't be in the RuntimeObject constructor because this is called before the renderer is created. In this case, still move this before the loop on the behaviors onCreated.
It's possible a behavior would interact with the effects, but they are not applied yet, which could in theory be a problem.

HarsimranVirk replied:
Done 👍

4ian commented on GDJS/Runtime/runtimeobject.ts:
```diff
@@ -219,6 +246,25 @@ namespace gdjs {
+     * @param newEffectData New effect data
+     * @returns
+     */
+    _hotReloadRuntimeObjectEffects(
```
This method is using only public methods (removeEffect, addEffect).
This means it could be done outside of the object, inside the hot reloader. So let's remove it from here and do it inside the hot reloader (because it will avoid cluttering gdjs.RuntimeObject, and avoid cluttering updateFromObjectData which would become ill named)

HarsimranVirk replied:
Done 👍

4ian commented on GDJS/Runtime/runtimeobject.ts:
```diff
@@ -232,8 +278,10 @@ namespace gdjs {
     ): boolean {
-      // If not redefined, mark by default the hot-reload as failed.
-      return false;
+      return this._hotReloadRuntimeObjectEffects(
```
It should not be the responsibility of `updateFromObjectData`, I think, to update the effects. The hot reload can do it, as it has all it needs (old/new effect data, and access to remove/addEffect)

4ian replied:
Actually I'm confused why you would even need this knowing that this is implemented already in the hot reloader. Remove these changes entirely.

HarsimranVirk replied:
I think this is dead code, I forgot to remove this (my mistake). I started with this before looking at the hot reloader, then I understood it should be a part of the hot reloader itself. I'll remove it in the next commit.

4ian commented on GDJS/Runtime/runtimeobject.ts:
```diff
@@ -730,6 +778,168 @@ namespace gdjs {
+      return this._rendererEffects;
+    }
+
+    initializeEffect(effectsData: EffectData) {
```
Why do we need both a `initializeEffect` and a `addEffect`? Can we have only one, `addEffect`?

HarsimranVirk replied:
The reason for breaking `addEffect` into multiple functions (`initializeEffect, applyEffect`) was because the renderer was created after the constructor (of runtime object) has been called. I wasn't sure how to retain the `effectsData` outside the constructor (and if it even made any sense to retain the initial effect data as a field in the object or not). Maybe there's a better design to handle it (instead of breaking the function and calling them at different times)?

4ian replied:
I see, we indeed need to `applyEffect` at the very end, outside of the base RuntimeObject constructor.
In this case: 
- add a comment when calling `initializeEffect` in the constructor ("We call initializeEffect as we can't use addEffect, as the renderer is not ready yet. This means applyEffect will have to be called later (see onCreated)").
- avoid this function entirely if you can (just do it in the constructor. Because each new function is complexifying the interface of the runtime object, which is already very large.
- if you can't avoid this function, at least make it private (underscore + private keyword in TypeScript). We don't want any outside caller to mess with this - it's an implementation detail, everyone else should be calling addEffect :)

4ian replied:
This being said, I can't think of a better design that would allow us to scrap entirely initialize/apply effect - not sure it's possible unfortunately!
Let's just try to remove as much initializeEffect as possible to keep the new interface very small.

4ian commented on GDJS/Runtime/runtimeobject.ts:
```diff
@@ -730,6 +778,168 @@ namespace gdjs {
+      this._runtimeScene
+        .getGame()
+        .getObjectEffectsManager()
+        .initializeEffect(effectsData, this._rendererEffects, layer);
```
We should not need to pass the layer?

HarsimranVirk replied:
Oh, this is definitely a mistake. But I've one doubt here, we definitely don't need a layer here, but an `EffectsTarget` (which is either a runtime object or a layer). So, we'll still need to pass the runtime object (instead of a layer), but I guess that should still be very expensive right?

4ian replied:
Well it depends :) What do you mean exactly by "this should still be very expensive"?
If we pass the RuntimeObject, i.e: `this`, it's a reference that is passed like all objects in JavaScript, so internally it's just basically passing a pointer as an argument, which is very cheap :) 

It seems fair to pass to something called "initializeEffect": the effects data, the map where to store effects and the target (i.e: the runtime object) (which can be queried for its width/height for example).

4ian commented on GDJS/Runtime/runtimeobject.ts:
```diff
@@ -730,6 +778,168 @@ namespace gdjs {
+          effectData,
+          this._rendererEffects,
+          this.getRendererObject(),
+          layer
```
Same here

4ian commented on GDJS/Runtime/pixi-renderers/pixi-object-effects-manager.ts:
```diff
@@ -0,0 +1,271 @@
+    ): boolean {
+      let effectAdded = true;
+      effectAdded =
+        this.initializeEffect(effectData, rendererEffects, layer) &&
```
Why passing layer here and not rendererObject?

4ian commented on GDJS/Runtime/pixi-renderers/pixi-object-effects-manager.ts:
```diff
@@ -0,0 +1,271 @@
+      effectData: EffectData,
+      rendererEffects: RendererEffects,
+      rendererObject: PIXI.DisplayObject,
+      layer: Layer
```
We should not pass the layer, the effects manager should just deal with `rendererObject`. Remove any mention to layer in this file.

4ian commented on GDJS/Runtime/pixi-renderers/pixi-object-effects-manager.ts:
```diff
@@ -0,0 +1,271 @@
+    /**
+     * Update the filters applied on an object.
+     * @param runtimeObject
+     * @param layer
+     */
+    update(rendererEffects: RendererEffects, layer: Layer) {
+      for (const filterName in rendererEffects) {
+        const filter = rendererEffects[filterName];
+        filter.update(filter.pixiFilter, layer);
```
No `layer` anywhere, just `rendererObject`

4ian commented on GDJS/Runtime/pixi-renderers/pixi-object-effects-manager.ts:
```diff
@@ -0,0 +1,271 @@
+  class PixiObjectEffectsManager {
+    /**
+     * Initialize the renderer effects from the effect data.
+     * @param effectData The effect data
```
Describe what is needed to do after the initialization (i.e: call applyEffect to actually add the effect to the renderer object).

4ian reviewed (commented):
We're approaching the solution, but we'll need to be extra precise on everything to ensure we have the absolutely simplest and straightforward implementation - every single line, parameter and method must be carefully inspected to ensure we've done the simplest thing without any extra layer/hot reloading/redundant method ;)

4ian commented on GDJS/Runtime/layer.ts:
```diff
@@ -463,10 +463,8 @@ namespace gdjs {
-      return (
-        this._runtimeScene.getTimeManager().getElapsedTime() * this._timeScale
-      );
+    getElapsedTime(runtimeScene: RuntimeScene): float {
+      return runtimeScene.getTimeManager().getElapsedTime() * this._timeScale;
```
Sorry for the u turn, I think we can actually do the reverse decision:
- keep getElapsedTime without any parameter in the layer, and remove the parameter from gdjs.RuntimeObject too.
- the reason is that both the layer and the runtimeobject have a reference to the runtimescene, and there is little to no chance that in the future we want to drop this reference.
- it's no problem if some code is passing the runtimescene as an argument to the object getElapsedTime, we can just ignore it. You can make the argument optional.

This means updating the EffectsTarget and also the filters using getElapsedTime - but hopefully won't be too much work :)

4ian commented on GDJS/Runtime/spriteruntimeobject.ts:
```diff
@@ -438,6 +438,7 @@ namespace gdjs {
      * Update the current frame of the object according to the elapsed time on the scene.
      */
     update(runtimeScene: gdjs.RuntimeScene): void {
+      super.update(runtimeScene);
```
You can remove this if you do the changes related to updating effects.

4ian commented:
Awesome, we're getting there! Take a look at GDJS tests that I think are failing. Rest should be ok, just see the comments :)

HarsimranVirk commented:
> Take a look at GDJS tests that I think are failing.

@4ian the tests are failing because of the firebase extension, they were failing since the time this PR was opened, and I assume it had been fixed in the later PRs. Should I pull the latest changes?

HarsimranVirk commented:
@4ian it turns out the tests are failing because they don't pass `effects` in the `objectData`. Should I add it everywhere? Or do something like `variables`, ie, only handle it if it exists?

Also, I think we've got duplicate commits after rebasing on master, let's open a new PR?

4ian commented:
> EDIT: I prefer adding effects everywhere, because they are more like behaviors, and not variables, it's going to be a little bit tedious to edit all of the tests, but this should be better I guess, what do you think?

Yes it's better, the `effects` array should be required (even if it's tedious to add first, it's more consistent and avoid surprises).

> Also, I think we've got duplicate commits after rebasing on master, let's open a new PR?

Yep!
